### DISCUSSION TITLE
Ideas for Saving Previous Queries

### DISCUSSION THREAD

**Discussion by BradyAJohnston** [ropensci/webchem#352] [General]
I am looking to try and incoroprate webchem into another package, for doing calculations for buffer solutions and other mass/molarity calculations. Currently works great for this, but I was wondering if any of the developers had any ideas for saving previous queries, so that the second time the user searched "potassium chloride" etc, you wouldn't have to query the online databases and could instead match it to a local set of previously used options. 

I have a couple of ideas of how I might implement it, but I was wondering if any of the developers had an idea that might use the webchem internals some more. 

Thanks for all the hard work in making this great package so far!

**Comment by Aariq** [comment]
I think I briefly considered trying to work on this using something like the [memoise package](https://cran.r-project.org/web/packages/memoise/index.html).  I think it's a great idea, but I think we'd need to devote some time looking into how to best implement it. I don't personally have any experience with memoisation or caching, so I'd need to do some learning to even know how feasible this is.

**Comment by Aariq** [comment]
Definitely interested in hearing your ideas though!

**Comment by stitam** [comment]
Thanks @BradyAJohnston for opening this discussion, I think implementing a way to quickly recover results would add a lot to the package!

I'm currently using the package for querying info about thousands of chemicals and what I currently do to make my life easier is cache the results using `save()`. I create a `.rda` file in my working directory and I use that file to quickly access the data afterwards. I use this method with loops, so when I query 1000 chemicals I cache the output after each query. This way when the function is interrupted for whatever reason I still have quick access to all the data that has already been retrieved. 

Thanks @Aariq for linking the memoise package, I think using a package that is already out there is probably better than finding our own way around this. Perhaps, we could have an argument in our functions where we specify whether we want to cache results or not and if we do we can cache them using the package?

**Comment by ethanbass** [comment]
I've been using memoise in conjunction with webchem (thanks to this post!). It seems like it might be kind of complicated to integrate memoisation into the package, but I wonder if it would be worth pointing out this possibility prominently in the documentation somewhere? It can really speed things up if you are trying run a list of queries where there a lot of repeats (and also reduce the number of calls to the server).

**Comment by stitam** [comment]
Thanks @ethanbass, I would prefer incorporating some kind of caching mechanism in the functions. The approach I described earlier was implemented in `chembl_query()`. Do you think implementing something like that for each function that interacts with databases could work?

**Reply by ethanbass** [reply]
I guess maybe it's not actually that complicated.  I think something like the function below could work if you wanted to use memoise for this. The cache argument here can be left as `NULL` (in which case the regular `foo` function is called) or it can be set to "memory" or "disk" to cache the results in memory or on the disk using memoise. If you want to also be able to specify the location of the cache on the disk, an additional argument could be added for this. (By default, `cachem::cache_disk` creates the cache file in a temporary directory if the `dir` isn't specified).

```
chembl_query <- function(query,
                         resource = "molecule",
                         cache = NULL,
                         verbose = getOption("verbose"),
                         test_service_down = FALSE) {
  resource <- match.arg(resource, chembl_resources())
  stem <- "https://www.ebi.ac.uk/chembl/api/data"
  foo <- function(query, verbose) {
    if (is.na(query)) {
      if (verbose) webchem_message("na")
      return(NA)
    }
    if (grepl("^CHEMBL[0-9]+", query) == FALSE) {
      if (verbose) message("Query is not a ChEMBL ID. Returning NA.")
      return(NA)
    }
    if (verbose) webchem_message("query", query, appendLF = FALSE)
    url <- ifelse(
      test_service_down, "", paste0(stem, "/", resource, "/", query, ".json"))
    webchem_sleep(type = "API")
    res <- try(httr::RETRY("GET",
                           url,
                           httr::user_agent(webchem_url()),
                           terminate_on = 404,
                           quiet = TRUE), silent = TRUE)
    if (inherits(res, "try-error")) {
      if (verbose) webchem_message("service_down")
      return(NA)
    }
    if (res$status_code != 200) {
      if (verbose) message(httr::message_for_status(res))
      return(NA)
    }
    if (verbose) message(httr::message_for_status(res))
    cont <- httr::content(res, type = "application/json")
    cont <- format_chembl(cont)
    return(cont)
  }
  if (!is.null(cache)){
    cache <- switch(cache, "memory" = cachem::cache_mem(),
           "disk" = cachem::cache_disk())
    foo <- memoise::memoise(foo, cache = cache)
  }
  out <- lapply(query, function(x) foo(x, verbose = verbose))
  return(out)
}
```

```
t1 <- system.time(results<-chembl_query(c("CHEMBL25", "CHEMBL25", "CHEMBL25"), cache="memory"))
t2 <- system.time(results<-chembl_query(c("CHEMBL25", "CHEMBL25", "CHEMBL25"), cache="disk"))
t3 <- system.time(results<-chembl_query(c("CHEMBL25", "CHEMBL25", "CHEMBL25"), resource = "molecule", cache=NULL))
```

(I'm not sure if there's a simple way to have the verbose output though with memoise, because it doesn't cache side effects like messages).

**Reply by Aariq** [reply]
I think it's recommended that you memoise functions on package load, not in functions themselves, which complicates things. https://memoise.r-lib.org/reference/memoise.html#details

>It is recommended that functions in a package are not memoised at build-time, but when the package is loaded. The simplest way to do this is within [.onLoad()](https://rdrr.io/r/base/ns-hooks.html) with, for example
>``` r
># file.R
>fun <- function() {
 >some_expensive_process()
>}
>```
> ```r
># zzz.R
>.onLoad <- function(libname, pkgname) {
> fun <<- memoise::memoise(fun)
>}
>```

**Reply by ethanbass** [reply]
Do we know what the reason is for this recommendation? (Technically, the memoisation in my function isn't done at build-time either, it is done each time the function is called).

**Reply by ethanbass** [reply]
One possible solution to this would be to bring the `foo` function outside of the "query" function, e.g. in this case: 
```
call_chembl <- function(query, verbose) {
    if (is.na(query)) {
      if (verbose) webchem_message("na")
      return(NA)
    }
    if (grepl("^CHEMBL[0-9]+", query) == FALSE) {
      if (verbose) message("Query is not a ChEMBL ID. Returning NA.")
      return(NA)
    }
    if (verbose) webchem_message("query", query, appendLF = FALSE)
    url <- ifelse(
      test_service_down, "", paste0(stem, "/", resource, "/", query, ".json"))
    webchem_sleep(type = "API")
    res <- try(httr::RETRY("GET",
                           url,
                           httr::user_agent(webchem_url()),
                           terminate_on = 404,
                           quiet = TRUE), silent = TRUE)
    if (inherits(res, "try-error")) {
      if (verbose) webchem_message("service_down")
      return(NA)
    }
    if (res$status_code != 200) {
      if (verbose) message(httr::message_for_status(res))
      return(NA)
    }
    if (verbose) message(httr::message_for_status(res))
    cont <- httr::content(res, type = "application/json")
    cont <- format_chembl(cont)
    return(cont)
  }

# zzz.R
.onLoad <- function(libname, pkgname) {
call_chembl_mem <<- memoise::memoise(call_chembl)
}

chembl_query <- function(query,
                         resource = "molecule",
                         cache = TRUE,
                         verbose = getOption("verbose"),
                         test_service_down = FALSE) {
  resource <- match.arg(resource, chembl_resources())
  stem <- "https://www.ebi.ac.uk/chembl/api/data"
  foo <- ifelse(cache, call_chembl_mem, call_chembl)
  out <- lapply(query, function(x) foo(x, verbose = verbose))
  return(out)
}
```

I'm not really sure what the reasoning is for this recommendation though...?

**Reply by ethanbass** [reply]
I did a little digging and from what I can tell, this recommendation seems to be mainly a way to circumvent an R CMD check warning that arises from using memoise at build-time. See https://github.com/r-lib/memoise/issues/76. I don't think my proposed function is likely to have this issue however, since memoise isn't actually run at buildtime, but only internally inside the function. 

(Edit: It doesn't seem like there is a warning. I guess one issue with this implementation is that the cache is not being saved between multiple calls to the function, which may not be the desired behavior. However, if you specify the path to a cache file on the disk (as in @stitam's original implementation), the results will seemingly be saved between multiple calls to the function).

**Reply by Aariq** [reply]
Thanks for tracking that down!  I never figured out _why_ that was a recommendation.

**Reply by Aariq** [reply]
I think I don't quite understand what is meant by "build-time dependency" here.  Why would memoising `foo()` inside an exported function not cause the same problem as memoising an exported function?

**Reply by ethanbass** [reply]
I don't claim to have a deep understanding of this, but I think the key distinction is between "build-time" and "run-time" dependencies. I think the warning is triggered by running memoise outside of a function, whereas if you run it inside of another function, it's a run-time dependency and there's no warning. 

I guess the distinction is based on whether you need the dependency to build the package. In the former case, you will not be able to install the package at all without memoise, whereas in the latter case, you can install the package, but you may not be able to run the function.

**Reply by ethanbass** [reply]
I'm having trouble reproducing the alleged R CMD check warning at all though, even when running memoise outside of a function. I made this little test package that I think should trigger the warning (https://github.com/ethanbass/memoise_test), but it doesn't.
